src/goto-programs/class_identifier.cpp

```diff
@@ -0,0 +1,88 @@
+/*******************************************************************\
+
+Module: Extract class identifier
+
+Author: Chris Smowton, chris.smowton@diffblue.com
+
+\*******************************************************************/
+
+#include "class_identifier.h"
+
+#include <util/std_expr.h>
+
+/*******************************************************************\
+
+Function: build_class_identifier
+
+  Inputs: Struct expression
+
+ Outputs: Member expression giving the clsid field of the input,
+          or its parent, grandparent, etc.
+
+ Purpose:
+
+\*******************************************************************/
+
+static exprt build_class_identifier(
+  const exprt &src,
+  const namespacet &ns)
+{
+  // the class identifier is in the root class
+  exprt e=src;
+
+  while(1)
+  {
+    const typet &type=ns.follow(e.type());
+    assert(type.id()==ID_struct);
+
+    const struct_typet &struct_type=to_struct_type(type);
+    const struct_typet::componentst &components=struct_type.components();
+    assert(!components.empty());
+
+    member_exprt member_expr(
+      e,
+      components.front().get_name(),
+      components.front().type());
+
+    if(components.front().get_name()=="@class_identifier")
+    {
+      // found it
+      return member_expr;
+    }
+    else
+    {
+      e=member_expr;
+    }
+  }
+}
+
+/*******************************************************************\
+
+Function: get_class_identifier_field
+
+  Inputs: Pointer expression of any pointer type, including void*,
+          and a recommended access type if the pointer is void-typed.
+
+ Outputs: Member expression to access a class identifier, as above.
+
+ Purpose:
+
+\*******************************************************************/
+
+exprt get_class_identifier_field(
+  exprt this_expr,
```

Comment:
I can see that you modify it below, but passing exprt by value is just uncommon and thus surprising.